pinot-common/src/main/java/org/apache/pinot/common/function/FunctionRegistry.java

```diff
@@ -94,83 +154,195 @@ private FunctionRegistry() {
    */
-  public static void registerFunction(String functionName, Method method, boolean nullableParameters,
-      boolean isPlaceholder, boolean isVarArg) {
-    if (!isPlaceholder) {
-      registerFunctionInfoMap(functionName, method, nullableParameters, isVarArg);
-    }
-    registerCalciteNamedFunctionMap(functionName, method, nullableParameters, isVarArg);
-  }
-
-  private static void registerFunctionInfoMap(String functionName, Method method, boolean nullableParameters,
-      boolean isVarArg) {
-    FunctionInfo functionInfo = new FunctionInfo(method, method.getDeclaringClass(), nullableParameters);
-    String canonicalName = canonicalize(functionName);
-    Map<Integer, FunctionInfo> functionInfoMap = FUNCTION_INFO_MAP.computeIfAbsent(canonicalName, k -> new HashMap<>());
-    if (isVarArg) {
-      FunctionInfo existFunctionInfo = functionInfoMap.put(VAR_ARG_KEY, functionInfo);
-      Preconditions.checkState(existFunctionInfo == null || existFunctionInfo.getMethod() == functionInfo.getMethod(),
-          "Function: %s with variable number of parameters is already registered", functionName);
+  @Nullable
+  public static FunctionInfo getFunctionInfo(SqlOperatorTable operatorTable, RelDataTypeFactory typeFactory,
+      String functionName, List<DataSchema.ColumnDataType> argTypes) {
+    PinotScalarFunction scalarFunction = getScalarFunction(operatorTable, typeFactory, functionName, argTypes);
+    if (scalarFunction != null) {
+      return scalarFunction.getFunctionInfo();
     } else {
-      FunctionInfo existFunctionInfo = functionInfoMap.put(method.getParameterCount(), functionInfo);
-      Preconditions.checkState(existFunctionInfo == null || existFunctionInfo.getMethod() == functionInfo.getMethod(),
-          "Function: %s with %s parameters is already registered", functionName, method.getParameterCount());
+      // TODO: convert this to throw IAE when all operator has scalar equivalent.
+      return null;
     }
   }
 
-  private static void registerCalciteNamedFunctionMap(String functionName, Method method, boolean nullableParameters,
-      boolean isVarArg) {
-    if (method.getAnnotation(Deprecated.class) == null) {
-      FUNCTION_MAP.put(functionName, ScalarFunctionImpl.create(method));
+  @Nullable
+  private static FunctionInfo getFunctionInfoFromCalciteNamedMap(String functionName, int numParams) {
+    List<PinotScalarFunction> candidates = getFunctionMap().range(functionName, CASE_SENSITIVITY).stream().filter(
+            e -> e.getValue() instanceof PinotScalarFunction && (e.getValue().getParameters().size() == numParams
+                || ((PinotScalarFunction) e.getValue()).isVarArgs())).map(e -> (PinotScalarFunction) e.getValue())
+        .collect(Collectors.toList());
+    if (candidates.size() == 1) {
+      return candidates.get(0).getFunctionInfo();
+    } else {
+      // TODO: convert this to throw IAE when all operator has scalar equivalent.
+      return null;
```

Comment:
Is this TODO going to be resolved after this PR is merged?

Reply:
yes but it will take quite a while. once we have every scalar function matches the transform we will throw here (e.g. we found a discrepancy between what transform can do but scalar can't)